### ISSUE TITLE
Multithreading for WL hasing function

### ISSUE THREAD

**Issue by zslwyuan** [benedekrozemberczki/karateclub#7]
Hi!

   Maybe just another suggestion. In the embedding algorithms, the WeisfeilerLehmanHashing function in the fit function could be time-consuming and the WL hashing function for each graph is independent. Therefore, maybe using multhreading from python can speed them up and I modify the code for my application of graph2vec:

==================================

    def fit(self, graphs):
        """
        Fitting a Graph2Vec model.

        Arg types:
            * **graphs** *(List of NetworkX graphs)* - The graphs to be embedded.
        """
        pool = ThreadPool(8)
        args_generator = [(graph, self.wl_iterations, self.attributed) for graph in graphs]
        documents = pool.starmap(WeisfeilerLehmanHashing, args_generator)
        pool.close()
        pool.join()
        #documents = [WeisfeilerLehmanHashing(graph, self.wl_iterations, self.attributed) for graph in graphs]
        documents = [TaggedDocument(words=doc.get_graph_features(), tags=[str(i)]) for i, doc in enumerate(documents)]

        model = Doc2Vec(documents,
                        vector_size=self.dimensions,
                        window=0,
                        min_count=self.min_count,
                        dm=0,
                        sample=self.down_sampling,
                        workers=self.workers,
                        epochs=self.epochs,
                        alpha=self.learning_rate,
                        seed=self.seed)

        self._embedding = [model.docvecs[str(i)] for i, _ in enumerate(documents)]

**Comment by benedekrozemberczki** [comment]
Hi there,

I will test this and if it works I will add it. You can also open a PR, but
that would require modifications in a large number of places. What do you
say?

Regards,

Benedek

**Comment by zslwyuan** [comment]
Hi Benedek,

Sure! After I double-check my project, I will open a PR. Thanks!

Best regards,

Tingyuan

**Comment by benedekrozemberczki** [comment]
I looked at it and it does not seem to improve the performance - the NX graphs are serialized and deserialized under the hood. This is generally a problem with NX and Python.

**Comment by zslwyuan** [comment]
Hi Benedek,

Yes, I noticed the situation and I am trying to make multithreading work properly for this application since it does need some acceleration for my dataset. ^_^ If I get progress, I will update.

Best regards,

Tingyuan

**Comment by benedekrozemberczki** [comment]
https://github.com/benedekrozemberczki/graph2vec ?
